design/open/hard-delete.md

```diff
@@ -0,0 +1,202 @@
+# Hard Delete Uncommitted Data
+
+## Motivation
+
+Uncommitted data which is no longer referenced (due to branch deletion, reset branch etc.) is not being deleted by lakeFS.
+This may result excessive storage usage and possible compliance issues.
+To solve this two approaches were suggested:
+1. A batch operation performed as part of an external process (GC)
+2. An online solution inside lakeFS
+
+**This document details the latter.**
+
+Several approaches were considered and described in this document. All proposal suffer from unresolved design issues, and
+so no specific approach was selected as a solution.
+
+### Base assumptions
+
+The following assumptions are required in all the suggested proposals:
+1. Copy operation requires to be changed and implemented is full copy.
+2. lakeFS is not responsible for retention of data outside the repository path (data ingest)
+3. An online solution is not bulletproof, and will require a complementary external solution to handle edge cases
+
+## 1. Write staged data directly to object store on a defined path
+
+This proposal suggests moving the staging area management from the ref store to the object store and defining a structured path for
+the object's physical address on the object store.
+
+### Performance degradation when using the object store
+
+Due to the basic concept of handling the staging data in the objects store, this proposal suffers from a significant performance 
+degradation in one of lakeFS's principal flows - **listing**. Listing is used in flows such as: list branch objects, getting objects and committing.
+We tested listing of ~240000 **staged objects** using `lakeFS ls` and comparing it with `aws s3 ls` with the following results:
+> ./lakectl fs ls lakefs://niro-bench/main/guy-repo/itai-export/medium/ >   6.42s user 0.78s system 44% cpu 16.089 total
+
+> aws s3 ls s3://guy-repo/itai-export/medium/ > /dev/null  56.82s user 2.34s system 74% cpu 1:19.79 total
+
+### Design
+
+Objects will be stored in a path relative to the branch and staging token. For example: file `x/y.z` will be uploaded to path
+`<bucket-name>/<repo-name>/<branch-name>/<staging-token>/x/y.z`.  
```

Comment:
I couldn't come up with any sufficient solution to this problem. The only option I thought of is adding a metadata entry on the object which marks it as deleted

Reply:
I think this is an important issue to resolve.  (I know it's hard because AFAIR we've had several rounds on it.)

Adding metadata doesn't work because then you need to get actual data for the object to know how to handle it.  E.g. think of listing, suddenly it has to HEAD every object in the listing.  So your *only* options are to use [data available in ListObjectsV2](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#API_ListObjectsV2_ResponseSyntax), where every object is represented by an [Object](https://docs.aws.amazon.com/AmazonS3/latest/API/API_Object.html).  AFAICT your only chance there is an ETag -- if you declared some sequence of bytes "The Deleted Object" then you could compute its ETag and try to upload _that_ whenever you wanted to delete an object.  I am **not a fan** of having fixed "forbidden" strings, however -- they're exactly the kind of thing people like me end up wanting to store inside actual objects.